File: java/src/main/java/com/authdog/resources/ProjectsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Project (application) management operations.
 */
public final class ProjectsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a projects helper.
     * @param clientParam Authdog client
     */
    public ProjectsResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * Create or update a project.
     * @param tenantId tenant ID
     * @param body request body
     * @return project envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode save(final String tenantId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                "/v1/tenants/" + tenantId + "/applications", body, null,
                JsonNode.class);
    }

    /**
     * Get a project and its environments.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @return project envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode get(final String tenantId,
                        final String applicationId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/applications/"
                        + applicationId,
                null, null, JsonNode.class);
    }

    /**
     * Delete a project.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String applicationId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                "/v1/tenants/" + tenantId + "/applications/"
                        + applicationId,
                null, null, JsonNode.class);
    }

    /**
     * Set a project's default environment.
     * @param tenantId tenant ID
     * @param applicationId application ID
     * @param body request body
     * @return update envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode setDefaultEnvironment(final String tenantId,
                                          final String applicationId,
                                          final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PUT",
                "/v1/tenants/" + tenantId + "/applications/"
                        + applicationId + "/default-environment",
                body, null, JsonNode.class);
    }
}

File: csharp/Resources/ProjectsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Project (application) Wave 1 operations
    /// </summary>
    public class ProjectsResource
    {
        private readonly AuthdogClient _client;

        public ProjectsResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> SaveAsync(string tenantId, object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, $"/v1/tenants/{tenantId}/applications", body);

        public JObject Save(string tenantId, object body) => SaveAsync(tenantId, body).GetAwaiter().GetResult();

        public Task<JObject> GetAsync(string tenantId, string applicationId) =>
            _client.RequestAsync<JObject>(HttpMethod.Get, $"/v1/tenants/{tenantId}/applications/{applicationId}");

        public JObject Get(string tenantId, string applicationId) =>
            GetAsync(tenantId, applicationId).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string applicationId) =>
            _client.RequestAsync<JObject>(HttpMethod.Delete, $"/v1/tenants/{tenantId}/applications/{applicationId}");

        public JObject Delete(string tenantId, string applicationId) =>
            DeleteAsync(tenantId, applicationId).GetAwaiter().GetResult();

        public Task<JObject> SetDefaultEnvironmentAsync(string tenantId, string applicationId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Put,
                $"/v1/tenants/{tenantId}/applications/{applicationId}/default-environment",
                body);

        public JObject SetDefaultEnvironment(string tenantId, string applicationId, object body) =>
            SetDefaultEnvironmentAsync(tenantId, applicationId, body).GetAwaiter().GetResult();
    }
}
